crates/agentgateway/src/proxy/proxy_protocol.rs

```diff
@@ -0,0 +1,222 @@
+//! PROXY protocol v2 parser for Istio sandwich waypoint mode.
+//!
+//! When agentgateway operates as an Istio ambient mesh waypoint in "sandwich" mode,
+//! ztunnel handles mTLS termination and forwards traffic to the waypoint using
+//! PROXY protocol v2. This module parses the PROXY header to extract:
+//!
+//! - Original source/destination addresses (standard PROXY protocol)
+//! - Peer identity from TLV 0xD0 (SPIFFE URI of the source workload)
+//!
+//! The extracted identity flows through to CEL authorization via TLSConnectionInfo.
+
+use std::net::SocketAddr;
+
+use anyhow::bail;
+use ppp::{HeaderResult, v2};
+use tokio::io::{AsyncRead, AsyncReadExt};
+use tracing::trace;
+
+use crate::strng;
+use crate::transport::tls::IstioIdentity;
+
+/// TLV type for peer identity (SPIFFE URI) - matches ztunnel's PROXY_PROTOCOL_AUTHORITY_TLV
+const PROXY_PROTOCOL_AUTHORITY_TLV: u8 = 0xD0;
+
+/// PROXY protocol v2 signature (12 bytes)
+const PROXY_V2_SIGNATURE: [u8; 12] = [
+	0x0D, 0x0A, 0x0D, 0x0A, 0x00, 0x0D, 0x0A, 0x51, 0x55, 0x49, 0x54, 0x0A,
+];
+
+/// Minimum header size: 12 (signature) + 4 (version/command/family/length)
+const PROXY_V2_MIN_HEADER: usize = 16;
+
+/// Information extracted from PROXY protocol v2 header.
+#[derive(Debug)]
+pub struct ProxyProtocolInfo {
+	/// Original source address of the client (before ztunnel)
+	pub src_addr: SocketAddr,
+	/// Original destination address (the service VIP)
+	pub dst_addr: SocketAddr,
+	/// Peer identity extracted from TLV 0xD0, if present
+	pub peer_identity: Option<IstioIdentity>,
+}
+
+/// Parse PROXY protocol v2 header from stream.
```

Comment:
I am fine with this for now, but in the future we will probably want...
* _optional_ PROXY; detect if its there or allow it to not be
* v1 support. I vaguely recall the `ppp` library has safety bugs with this (either with v1 in general, or detecting v1 vs v2) so we should be cautious

But fine to be followups for sure